crates/conjure_core/src/rules/base.rs

```diff
@@ -79,59 +79,33 @@ fn remove_nothings(expr: &Expr, _: &Model) -> ApplicationResult {
     }
 }
 
-/**
- * Remove empty expressions:
- * ```text
- * [] = Nothing
- * ```
- */
-#[register_rule(("Base", 100))]
-fn empty_to_nothing(expr: &Expr, _: &Model) -> ApplicationResult {
-    match expr {
-        Expr::Nothing | Expr::Reference(_, _) | Expr::Constant(_, _) => {
-            Err(ApplicationError::RuleNotApplicable)
-        }
-        _ => {
-            if expr.children().is_empty() {
-                Ok(Reduction::pure(Expr::Nothing))
-            } else {
-                Err(ApplicationError::RuleNotApplicable)
-            }
-        }
+/// Remove empty expressions:
+///  ```text
+///   or([]) ~> false
+///   X([])  ~> Nothing
+///  ```
+#[register_rule(("Base", 8800))]
+fn remove_empty_expression(expr: &Expr, _: &Model) -> ApplicationResult {
+    use Expr::*;
+
+    // excluded expressions
+    if matches!(
+        expr,
+        Nothing | Reference(_, _) | Constant(_, _) | WatchedLiteral(_, _, _)
+    ) {
+        return Err(ApplicationError::RuleNotApplicable);
     }
-}
 
-/**
- * Evaluate sum of constants:
- * ```text
- * sum([1, 2, 3]) = 6
- * ```
- */
-#[register_rule(("Base", 100))]
-fn sum_constants(expr: &Expr, _: &Model) -> ApplicationResult {
-    match expr {
-        Expr::Sum(_, exprs) => {
-            let mut sum = 0;
-            let mut new_exprs = Vec::new();
-            let mut changed = false;
-            for e in exprs {
-                match e {
-                    Expr::Constant(_metadata, Const::Int(i)) => {
-                        sum += i;
-                        changed = true;
-                    }
-                    _ => new_exprs.push(e.clone()),
-                }
-            }
-            if !changed {
-                return Err(ApplicationError::RuleNotApplicable);
-            }
-            // TODO (kf77): Get existing metadata instead of creating a new one
-            new_exprs.push(Expr::Constant(Metadata::new(), Const::Int(sum)));
-            Ok(Reduction::pure(Expr::Sum(Metadata::new(), new_exprs))) // Let other rules handle only one Expr being contained in the sum
-        }
-        _ => Err(ApplicationError::RuleNotApplicable),
+    if !expr.children().is_empty() {
+        return Err(ApplicationError::RuleNotApplicable);
     }
+
+    let new_expr = match expr {
+        Or(_, _) => Constant(Metadata::new(), Const::Bool(false)),
+        _ => Nothing,
```

Comment:
Out of interest, what other types of things can appear here?

Reply:
This generically covers all the vector operations Sum[], And[], Or[], Mul[], etc